lib/invoker/dns_cache.rb

```diff
@@ -2,22 +2,84 @@ module Invoker
   class DNSCache
     attr_accessor :dns_data
 
+    # INI:
+    # [app]
+    # port = 8000
+    # location = www2.app/blah
+    #
+    # [chat]
+    # port = 8001
+    # location www.app/chat
+    #
+    # [api]
+    # port = 8002
+    # location = www.app/api
+    #
+    # dns_data:
+    # {
+    #   'api' => [
+    #     ['', 8002]
+    #   ],
+    #   'app' => [
+    #     ['', 8000]
+    #   ],
+    #   'chat' => [
+    #     ['', 8001]
+    #   ],
+    #   'www.app' => [
+    #     ['', 8000],
+    #     ['/api, 8002],
+    #     ['/chat, 8001]
+    #   ],
+    #   'www2.app' => [
+    #     ['/blah', 8000]
+    #   ]
+    # }
     def initialize(config)
-      self.dns_data = {}
+      self.dns_data = Hash.new {|h,k| h[k] = []}
       @dns_mutex = Mutex.new
       Invoker.config.processes.each do |process|
         if process.port
-          dns_data[process.label] = { 'port' => process.port }
+          add(process.label, process.port)
+          if process.location
+            process.location.split(' ').each do |loc|
+              add(loc, process.port)
+            end
+          end
         end
       end
     end
 
-    def [](process_name)
-      @dns_mutex.synchronize { dns_data[process_name] }
+    def find_process(host, path)
+      @dns_mutex.synchronize {
+        until dns_data.include?(host) || host.nil?
+          host = host.split('.', 2)[1]
+        end
+
+        dns_data[host].reverse_each {|prefix, port|
+          if path_matches_prefix?(path, prefix)
+            return {:port => port}
+          end
+        }
+
+        nil
+      }
+    end
+
+    def add(location, port)
+      @dns_mutex.synchronize {
+        host, path_prefix = split_host_path(location)
+        (dns_data[host] << [path_prefix.to_s, port]).sort_by! &:length
+      }
+    end
+
+    private
+    def split_host_path(label)
+      label.split(%r{(?=/)}, 2)
```

Comment:
The `path` only refers to the section following the domain the URL, whereas location refers to the app name + `path`.

Although we are not explicitly adding path support to process labels, it's actually easier to support path than not, since both process labels and locations go through the same location parser. It also makes sense, in my opinion, since there's no clear way to *not* support paths in labels. We could, perhaps, add to code to error out, or add code to ignore any process labels that have paths, but why not just let it be handled in the same way location is handled?